src/uu/more/src/more.rs

```diff
@@ -264,54 +270,52 @@ struct Pager<'a> {
     // The current line at the top of the screen
     upper_mark: usize,
     // The number of rows that fit on the screen
-    content_rows: usize,
+    content_rows: u16,
     lines: Vec<String>,
     next_file: Option<&'a str>,
     line_count: usize,
-    close_on_down: bool,
     silent: bool,
 }
 
 impl<'a> Pager<'a> {
-    fn new(rows: usize, lines: Vec<String>, next_file: Option<&'a str>, silent: bool) -> Self {
+    fn new(rows: u16, lines: Vec<String>, next_file: Option<&'a str>, silent: bool) -> Self {
         let line_count = lines.len();
         Self {
             upper_mark: 0,
-            content_rows: rows - 1,
+            content_rows: rows.saturating_sub(1),
             lines,
             next_file,
             line_count,
-            close_on_down: false,
             silent,
         }
     }
 
     fn should_close(&mut self) -> bool {
-        if self.upper_mark + self.content_rows >= self.line_count {
-            if self.close_on_down {
-                return true;
-            }
-            if self.next_file.is_none() {
-                return true;
-            } else {
-                self.close_on_down = true;
-            }
-        } else {
-            self.close_on_down = false;
-        }
-        false
+        self.upper_mark
+            .saturating_add(self.content_rows.into())
+            .eq(&self.line_count)
     }
 
     fn page_down(&mut self) {
-        self.upper_mark += self.content_rows;
+        self.upper_mark = self
+            .upper_mark
+            .saturating_add(self.content_rows.into())
+            .min(self.line_count.saturating_sub(self.content_rows.into()));
```

Comment:
I implemented something like this before, but the simple version just seemed to work fine. Were you encountering any issues with it? I'm thinking maybe it doesn't scroll up a full page after getting to the end in my version?

Reply:
Need this bound so that the new `should_close` implementation works. But I think I can remove the need by using `gte` instead of `eq`.

Reply:
I think you can always add `content_rows` and change the check in `should_close` to `>` instead of `eq`.

Reply:
I removed the additional `min` here https://github.com/uutils/coreutils/pull/2407/commits/083e74597613d2edd1d17fce1bca282a8db0f315

It should still work fine after I changed `eq` to `gte` like @miDeb mentioned.